### PR TITLE
[MFMA] Move operand casts to AccelerateMatMul pass

### PR DESCRIPTION
**Pull request by binarman**
This PR moves casts of operands from python code and ttg to llvm phase to AccelerateAMDMatmul pass

### CODE REVIEW

File: python/triton/language/semantic.py [hunk 1]
```diff
@@ -1352,11 +1333,11 @@ def assert_dtypes_valid(lhs_dtype, rhs_dtype, target):
...
     # max_num_imprecise_acc only applies to fp8 -> fp32 dot on sm_90
-    if not (_is_cuda(builder.target) and builder.target.capability == 90 and lhs.dtype.is_fp8() and rhs.dtype.is_fp8()
-            and ret_scalar_ty.is_fp32()):
-        max_num_imprecise_acc = 0
     if max_num_imprecise_acc is None:
-        max_num_imprecise_acc = 2**30
+        if lhs.dtype.is_fp8() and rhs.dtype.is_fp8():
+            max_num_imprecise_acc = builder.options.max_num_imprecise_acc_default
+        else:
+            max_num_imprecise_acc = 0
```
Comment (binarman) [inline python/triton/language/semantic.py:1340]: I've just copied this part from upstream to make this function more upstream-like.
Currently we do not run this code, so nothing should change for us.

**Comment by alefimov-amd** [comment]
Converted to draft, because first want to merge mfma4x64 support, and tests generating scripts (for accelerate matmul lit tests)

**Comment by zhanglx13** [comment]
@alefimov-amd @scxiao 
I'd suggest we close this one.  The type promotion is upstream'ed and we don't need it in terms of perf.

**Comment by alefimov-amd** [comment]
@zhanglx13 
I want to update this PR, when upstream is stable(at least one more PR: https://github.com/openai/triton/pull/3025), to keep upstream and triton-mlir compatible.

Eventually, we will need to transfer all the stuff we have in triton-mlir, if upstream and triton-mlir are compatible, we can simply copy code from triton-mlir to upstream.

**Comment by zhanglx13** [comment]
@alefimov-amd  Ok, I'm also fine with that

**Comment by binarman** [comment]
This PR did 3 things:

1, Move C/D operand casts from python to C++ code
2. Remove redundant MFMA support checks (we do not need them, since we have new instruction selection)
3. Add casts of A/B operands in case of mixed dtype inputs

Now this PR is doing only 1 item.
2 item is moved to #496
3 item is temporary removed, since mixed precision is not supported in tt.dot operation(MLIR verification fails),
I want to discuss  this with upstream first

File: lib/Dialect/TritonGPU/Transforms/AccelerateAMDMatmul.cpp [hunk 2]
```diff
@@ -284,12 +358,16 @@ class BlockedToMFMA : public mlir::RewritePattern {
...
+                                             dotOp.getMaxNumImpreciseAcc());
+
+    Value dotOutput =
+        convertAndCastTensor(rewriter, newDot, oldRetType.getEncoding(),
+                             oldRetType.getElementType());
+
+    rewriter.replaceOp(op, dotOutput);
 
-    rewriter.replaceOpWithNewOp<ttg::ConvertLayoutOp>(op, oldRetType,
-                                                      newDot.getResult());
```
Comment (binarman) [inline lib/Dialect/TritonGPU/Transforms/AccelerateAMDMatmul.cpp:292]: changes in this file are copied from upstream PRs: https://github.com/openai/triton/pull/3002 and https://github.com/openai/triton/pull/3025

**Pull request merged**